File: Infra/Infra/Semaphore.cpp
#include "Semaphore.hpp"




CppClassNew(Semaphore)






Bool Semaphore_Init(Int o)
{
    Semaphore* m;

    m = CP(o);




    Int initCount;

    initCount = m->InitCount;




    int ua;

    ua = initCount;




    m->Intern = new QSemaphore(ua);




    return true;
}






Bool Semaphore_Final(Int o)
{
    Semaphore* m;

    m = CP(o);



    delete m->Intern;




    return true;
}

CppField(Semaphore, InitCount)

Int Semaphore_CountGet(Int o)
{
    Semaphore* m;

    m = CP(o);



    int u;

    u = m->Intern->available();



    Int a;

    a = u;


    return a;
}





Bool Semaphore_CountSet(Int o, Int value)
{
    return true;
}





Bool Semaphore_Acquire(Int o)
{
    Semaphore* m;

    m = CP(o);



    m->Intern->acquire();



    return true;
}





Bool Semaphore_Release(Int o)
{
    Semaphore* m;

    m = CP(o);



    m->Intern->release();



    return true;
}

File: Avalon/Avalon.Thread/Semaphore.cs
namespace Avalon.Thread;

public class Semaphore : Any
{
    public override bool Init()
    {
        base.Init();
        ulong ua;
        ua = (ulong)(this.InitCount);
        this.Intern = Extern.Semaphore_New();
        Extern.Semaphore_InitCountSet(this.Intern, ua);
        Extern.Semaphore_Init(this.Intern);
        return true;
    }

    public virtual bool Final()
    {
        Extern.Semaphore_Final(this.Intern);
        Extern.Semaphore_Delete(this.Intern);
        return true;
    }

    public virtual int InitCount { get; set; }
    private ulong Intern { get; set; }

    public virtual bool Acquire()
    {
        Extern.Semaphore_Acquire(this.Intern);
        return true;
    }

    public virtual bool Release()
    {
        Extern.Semaphore_Release(this.Intern);
        return true;
    }

    public virtual int Count
    {
        get
        {
            ulong u;
            u = Extern.Semaphore_CountGet(this.Intern);
            int a;
            a = (int)u;
            return a;
        }
        set
        {
        }
    }
}